Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -301,163 +323,173 @@ extension AnyMessageStorage {
-    if case .binary(let myValue) = state, case .binary(let otherValue) = other.state, myValue == otherValue {
-      return true
-    }
-
-    // If both have contentJSON, and they exactly match; the messages are equal.
-    // Because there could be map in the message (or the JSON could just be in a different
-    // order), the fact that the JSON isn't the same doesn't always mean the messages
-    // aren't equal.
-    if case .contentJSON(let myJSON, _) = state,
-       case .contentJSON(let otherJSON, _) = other.state,
-       myJSON == otherJSON {
-      return true
-    }
-
-    // Out of options. To do more compares, the states conversions would have to be
-    // done to do comparisions; and since equality can be used somewhat removed from
-    // a developer (if they put protos in a Set, use them as keys to a Dictionary, etc),
-    // the conversion cost might be to high for those uses.  Give up and say they aren't equal.
-    return false
-  }
+      #if swift(>=4.2)
+        // Can't use _valueData for a few reasons:
+        // 1. Since decode is done on demand, two objects could be equal
+        //    but created differently (one from JSON, one for Message, etc.),
+        //    and the hash values have to be equal even if we don't have data
+        //    yet.
+        // 2. map<> serialization order is undefined. At the time of writing
+        //    the Swift, Objective-C, and Go runtimes all tend to have random
+        //    orders, so the messages could be identical, but in binary form
+        //    they could differ.
+        public func hash(into hasher: inout Hasher) {
+          if !_typeURL.isEmpty {
+            hasher.combine(_typeURL)
+          }
+        }
+      #else  // swift(>=4.2)
+        var hashValue: Int {
+          var hash: Int = i_2166136261
+          if !_typeURL.isEmpty {
+            hash = (hash &* i_16777619) ^ _typeURL.hashValue
+          }
+          return hash
+        }
+      #endif
+
+      func isEqualTo(other: AnyMessageStorage) -> Bool {
+            if _typeURL != other._typeURL { return false }
+
+            // Since the library does lazy Any decode, equality is a very hard problem.
+            // It things exactly match, that's pretty easy, otherwise, one ends up having
+            // to error on saying they aren't equal.
+            //
+            // The best option would be to have Message forms and compare those, as that
+            // removes issues like map<> serialization order, some other protocol buffer
+            // implementation details/bugs around serialized form order, etc.; but that
+            // would also greatly slow down equality tests.
+            //
+            // Do our best to compare what is present have...
+
+            // If both have messages, check if they are the same.
+            if case .message(let myMsg) = state,
+                  case .message(let otherMsg) = other.state, type(of: myMsg)
+                  == type(of: otherMsg) {
```

Comment:
Drive by - Scanning this file, this wrap on the `==` took me a moment to figure out what the line really was.  Not sure if there is a better thing to be done here.

Reply:
We should put groups around each of the comma-separated clauses to prefer that they wrap as a whole instead of breaking in the middle of one of the expressions unless absolutely necessary.

The curly brace should have been moved to the next line; that's a minor bug I'll have to track down.

Reply:
@allevato: +1.  That said, if it *was* going break before `==` (as will sometimes be unavoidable) it's arguable that the `==` should be further indented